### ISSUE TITLE
PSR Http client and redirects

### ISSUE THREAD

**Issue by vaclavvanik** [guzzle/guzzle#2584]
[SendRequest method in Client.php](https://github.com/guzzle/guzzle/blob/192d1f1945e63239b45abfcfd606556f1673ab79/src/Client.php#L123)  disables redirects.

Is there any reason why? This behavior is hardcoded and cannot be overriden. As I know [PSR HTTP client](https://www.php-fig.org/psr/psr-18/) does not prohibit redirects.

Current behavior should be removed until stable version released.

**Comment by gmponos** [comment]
Yes.. in a way it does not clearly states it... but the PSR mentions

> A second goal is that HTTP clients can be replaced as per the Liskov substitution principle. This means that all clients MUST behave in the same way when sending a request.

So the point is that a PSR HTTP client must always return the response AS IS and the consumer should choose how to handle the response.

if we allowed or opt-in the redirects then we could never fulfill the above PSR statement.

**Comment by vaclavvanik** [comment]
Well it is not easy to make a decision. You wrote `client must always return the response AS IS`. This means every middleware which changes response (eg deflate gzipped response body) has to be disabled.

Personally I would allow redirects. If anyone wants to turn it off, it could be done in constructor config.

**Comment by gmponos** [comment]
> client must always return the response AS IS

OK let me rephrase it/extend it a bit..
What I meant is that if you have this kind of code:

```
$this->psrClient->sendRequest(new Request('GET', 'www.UrlThatRedirects.com') === 301;
```

The code above works depending on the configuration of the client... follow redirects or not... Which breaks the LSP.

BTW there are many useful things to be read also in the metadocument of the PSR-18

 - https://www.php-fig.org/psr/psr-18/meta/#default-behavior
 - https://www.php-fig.org/psr/psr-18/meta/#middleware-and-wrapping-a-client

Also the discussions that lead to the PSR-18 can be found here.. you might find more useful things.

https://github.com/php-http/fig-standards/issues

**Comment by vaclavvanik** [comment]
I understand what you mean. I wrote that with your point of view all middlewares which can modify response have to be disabled too.

Example:

[Decode content](http://docs.guzzlephp.org/en/stable/request-options.html#decode-content) option is not disabled in `sendRequest` method.

Then this code will not work:

```php
$response = $psrClient->sendRequest(new Request('GET', 'http://www.example.com/foo.gz');
$body = gzinflate((string) $response->getBody());
```

Btw. `This means that all clients MUST behave in the same way when sending a request.`
That may mean response cannot be modified too (eg adding Auth header from constructor config).

I suggest that the current redirect behavior remain same - `opt-in`.

**Comment by gmponos** [comment]
Ha.. I knew I have seen somewhere about the gzip thing.. https://github.com/php-http/fig-standards/issues/37

Not sure what answers you can get from that discussion though.. IMHO I quite agree with this comment https://github.com/php-http/fig-standards/issues/37#issuecomment-328107526

**Comment by Nyholm** [comment]
The default http client should not follow redirects. However, if you as an application author adds middleware to "follow redirects" you have intentionally broken the contract and you have to suffer the consequences (or be happy that it just works). 

> I suggest that the current redirect behavior remain same - opt-in.

That is not possible, that would mean that Guzzle is useless as PSR-18 out of the box.
